pkg/ruler/ruler.go

```diff
@@ -150,58 +140,53 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+//|                                                               |
+//|                   Query       +-------------+                 |
+//|            +------------------>             |                 |
+//|            |                  |    Store    |                 |
+//|            | +----------------+             |                 |
+//|            | |     Rules      +-------------+                 |
+//|            | |                                                |
+//|            | |                                                |
+//|            | |                                                |
+//|       +----+-v----+   Filter  +------------+                  |
+//|       |           +----------->            |                  |
+//|       |   Ruler   |           |    Ring    |                  |
+//|       |           <-----------+            |                  |
+//|       +-------+---+   Rules   +------------+                  |
+//|               |                                               |
+//|               |                                               |
+//|               |                                               |
+//|               |    Load      +-----------------+              |
+//|               +-------------->                 |              |
+//|                              |     Manager     |              |
+//|                              |                 |              |
+//|                              +-----------------+              |
+//|                                                               |
+//+---------------------------------------------------------------+
 type Ruler struct {
 	services.Service
 
-	cfg            Config
-	notifierCfg    *config.Config
-	managerFactory ManagerFactory
-
+	cfg         Config
 	lifecycler  *ring.BasicLifecycler
 	ring        *ring.Ring
 	subservices *services.Manager
-
-	store  rules.RuleStore
-	mapper *mapper
-
-	// Structs for holding per-user Prometheus rules Managers
-	// and a corresponding metrics struct
-	userManagerMtx     sync.Mutex
-	userManagers       map[string]*promRules.Manager
-	userManagerMetrics *ManagerMetrics
-
-	// Per-user notifiers with separate queues.
-	notifiersMtx sync.Mutex
-	notifiers    map[string]*rulerNotifier
+	store       rules.RuleStore
+	manager     MultiTenantManager
 
 	registry prometheus.Registerer
 	logger   log.Logger
 }
 
 // NewRuler creates a new ruler from a distributor and chunk store.
-func NewRuler(cfg Config, managerFactory ManagerFactory, reg prometheus.Registerer, logger log.Logger, ruleStore rules.RuleStore) (*Ruler, error) {
-	ncfg, err := buildNotifierConfig(&cfg)
-	if err != nil {
-		return nil, err
-	}
-
-	userManagerMetrics := NewManagerMetrics()
-
-	if reg != nil {
-		reg.MustRegister(userManagerMetrics)
-	}
-
+func NewRuler(cfg Config, manager MultiTenantManager, reg prometheus.Registerer, logger log.Logger, ruleStore rules.RuleStore) (*Ruler, error) {
```

Comment:
@pracucci - I've made one more small change after you reviewed this PR (Sorry about this, I was going to create another follow up PR but unintentionally pushed changes as part of this itself 🤦 ). The change is that `NewRuler` now accepts a `MultiTenantManager` instead of a `ManagerFactory`.

Reply:
It's OK. It's part of this refactoring 👍